client/jetpack-cloud/sections/manage/pricing/license-products-list/featured-license-multi-item-card.tsx

```diff
@@ -119,7 +117,7 @@ export const FeaturedLicenseMultiItemCard = ( {
 	);
 
 	const onSelectProduct = useCallback( () => {
-		page( getIssueLicenseURL( variantSlug, bundleSize ) );
+		page.redirect( getIssueLicenseURL( variantSlug, bundleSize ) );
```

Comment:
For this purpose, it's better to call `page.replace`. It does exactly the same thing, but avoids an area where the page.js API is messy:
- `page( '/foo', handlerFn )` defines a route
- `page.redirect( '/foo', '/bar' )` defines a route that redirects to another route
- `page.show( '/foo' )` navigates to a route
- `page.replace( '/foo' )` navigates to a route while replacing the last history item with the new one

The messy part is that `page( '/foo' )` is the same as `page.show( '/foo' )` and `page.redirect( '/foo' )` is the same as `page.replace( '/foo' )`. The "route declaration" and "navigation command" functions are not separate, but are mixed together. Depending on the number and type (string or function) of the arguments.

Another special case for `page.redirect` is redirecting inside a route handler:
```js
page( '/foo', ( ctx ) => {
  if ( shouldBar ) {
    page.redirect( '/bar' ); // `ctx.page.redirect( '/bar' )` is even better
  } else {
    showFoo();
  }
} );
```
Here it's the same thing as `page.replace( '/bar' )` (i.e., not really a special case) but the convention when inside a route handler is to call it `page.redirect` although it's a command, not route a definition. Probably because of the analogy with a server router that would send a HTTP redirect response.

Reply:
@jsnajdr to clarify, though, which should be used when one clicks on a button to navigate to a new page? I'd expect `page.show()`, as we're navigating to a route after a click, but unfortunately this doesn't work, because it doesn't push the new navigator URL prior to checking for auth (the reason for this workaround in the first place). Would this be considered a bug?

Reply:
One difference between `page.show` and `page.replace` is that they do certain things in different order.

- [`page.show`](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-router/src/index.js#L576-L581) first does a "dispatch", i.e., finds the matching route handlers and executes them. And only after the dispatch does it call `history.pushState`, changing the URL.
- [`page.replace`](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-router/src/index.js#L649-L652) does this in the opposite order. First it calls `history.replaceState` and only then it does dispatch.

If the route handlers are synchronous and if they read `window.location`, they will see different values.

When your code is "checking for auth", does that involve reading `window.location` in a route handler?

Is this a bug? I'd say yes. I don't know why page.js behaves this way, and I'm curious what other, more modern routers are doing in this situation. This behavior is also a bit unstable, in the sense that sync and async handlers will see different values. Consider this:
```js
page( '/foo', ( ctx ) => {
  console.log( 'sync', window.location.href );
  Promise.resolve().then( () => console.log( 'async', window.location.href ) );
} );
```
If you call `page.show( '/foo' )`, then both console messages will report different locations! One will be the prev-location, the other the next-location. That's weird and very difficult to understand.

Reply:
> One difference between page.show and page.replace is that they do certain things in different order.

Indeed; earlier I was looking if this had changed, but `page.js` seems to be abandoned:
https://github.com/visionmedia/page.js/blob/master/index.js

> When your code is "checking for auth", does that involve reading `window.location` in a route handler?

This was happening in `authorizePath()` itself; I proposed a very narrow fix that would avoid this scenario by looking at the page context instead of the `window.location` in #87198 but it was decided to go with the route in this PR.

I still think it'd be wise to patch `authorizePath` to not be so reliant on `window.location`.

Reply:
FYI, I updated `page.redirect` with `page.replace` in #87287.